blog/2020-10-01-hs-to-coq/2020-10-01-hs-to-coq.md

```diff
@@ -0,0 +1,1084 @@
+are a few nitty-gritty details to take care of.
+
+We need some settings to tell the Coq interactive environment
+where to find *hs-to-coq* dependencies.
+
+<div class="tuto">
+Create a `_CoqProject` file at the root of the directory, with the following contents:
+
+```
+-R hs-to-coq/base ""
+-Q hs-to-coq/base-thy Proofs
+-Q src-coq Src
+-Q theories Proving
+```
+</div>
+
+#### Verification: Specification and Proof
+
+<div class="tuto">
+Using an editor configured for Coq,
+open a new file `theories/QueueSpec.v`.
+
+```
+emacs theories/QueueSpec.v
+```
+</div>
+
+##### Import dependencies
+
+In the first few lines, we load the required dependencies.
+
+- The `List` module from the standard library contains general facts about
+  lists;
+- The `GHC.List` module from *hs-to-coq* contains facts about Coq-ified
+  functions from the Haskell standard library (we will only need one fact
+  relating Haskell's `reverse` to Coq's `rev` function);
+- The `Queue` module we just translated and compiled.
+
+<div class="tuto">
+```alectryon
+From Coq Require Import List.
+From Proofs Require Import GHC.List.
+From Src Require Import Queue.
+```
+</div>
+
+The following `Import` command makes some extra *notations* available.
+Notations are user-defined syntactic sugar; in particular, infix operators
+are defined as notations in Coq.
+The module `ListNotations` contains the square bracket notations
+for lists: `[]`, `[x]`, `[x;y]`, `[x;y;z]`, etc.
+Two other notations for lists, `::` and `++`, are already available by default.
+
+<div class="tuto">
+```alectryon
+Import ListNotations.
+```
+</div>
+
+##### Define the model
+
+A queue can be modelled as a plain `list` of the values it contains.
+A list `xs` *models*---is a model for---a queue `q = MkQueue back front` when `xs` is equal to the
+concatenation of the `front` and the reversed `back` of the queue.
+
+```coq
+xs = front ++ rev back
+```
+
+This conformance relation between the queue and its model in Coq is a
+*predicate* on lists and queues;
+in code, the proposition "`xs` models `q`" will be written in Coq as `models xs q`.
+
+Formally, the proposition `models xs q` is a type with one constructor,[^def]
+`models_MkQueue`, which constructs a proof of
+`models xs (MkQueue back front)` from a proof of the proposition
+`xs = front ++ rev back`.
+This is the only way to construct a proof of a `models` proposition,
+and given a proof we can in fact *pattern-match* on it to extract the contained
+proof.
+Thanks to this constructor and pattern-matching,
+the following definition of `models` should be read to mean that
+the proposition `models xs (MkQueue back front)` is equivalent to
+`xs = front ++ rev back`.
+
+[^def]: We could also define `models xs q` to be *equal* to the proposition
+  `exists front back, q = MkQueue front back /\ xs = front ++ rev back`,
+  rather than wrap it an inductive type.
+  But definitions can be inconveniently unfolded,
+  which is why we prefer a new inductive type.
```

Comment:
Why does this need to be an inductive type? Wouldn't the following work?
```coq
Definition models {a} (xs : list a) (q : Queue a) : Prop :=
  forall front back, xs = front ++ rev back.
```

Reply:
That would work, although it makes things a little difficult (for the level of the tutorial) because we would have to explain how to provide the arguments of a `models` hypothesis.

In the end I found something even simpler than the inductive, a `Definition` using projections instead of quantifiers:

    Definition models xs q : Prop := front q ++ rev (back q) = xs.

Reply:
That's much better indeed.